### PR TITLE
Add support for SPV_KHR_ray_tracing

### PR DESCRIPTION
**Pull request by XAMPPRocky**
This PR adds initial support for ray tracing in Rust-GPU.

~~Currently the tests don't pass because I ran into some ICE's when testing `OpTerminateRayKHR`. Filed as https://github.com/EmbarkStudios/rust-gpu/issues/564~~

### CODE REVIEW

**Comment by XAMPPRocky** [comment]
cc @hrydgard @repi

**Comment by repi** [comment]
Cool!

**Comment by hrydgard** [comment]
Very cool!

To actually do real shading once we hit stuff with rays though, we need to be able to look up mesh data by mesh ID, which will likely end up requiring some of https://github.com/EmbarkStudios/rust-gpu/pull/450 .

File: crates/rustc_codegen_spirv/src/symbols.rs
Comment (hrydgard) [inline crates/rustc_codegen_spirv/src/symbols.rs:134]: khr on left side, NV on right? were some NV names reused in the KHR extension?

Comment (XAMPPRocky) [reply crates/rustc_codegen_spirv/src/symbols.rs:134]: It's just because rspirv has the NV versions but not the KHR versions.

Comment (expenses) [reply crates/rustc_codegen_spirv/src/symbols.rs:134]: Probably best to wait for an rspirv update then. The NV things are _probably_ the same as the KHR ones but it's best no to mix them up. Looking foward to this though!

Comment (khyperia) [reply crates/rustc_codegen_spirv/src/symbols.rs:134]: Huh? rspirv has the KHR versions already though https://docs.rs/spirv_headers/1.5.0/spirv_headers/enum.BuiltIn.html#associatedconstant.LaunchIdKHR

Comment (XAMPPRocky) [reply crates/rustc_codegen_spirv/src/symbols.rs]: Ah, I figured it out. They're not variants but constants. So you can't use `use Builtin::*;` and then say `LaunchIdKHR` because it's not imported with the variants. I've updated the PR to use those constants.

File: crates/rustc_codegen_spirv/src/symbols.rs
Comment (hrydgard) [inline crates/rustc_codegen_spirv/src/symbols.rs:146]: Shouldn't we just get rid of the NV raytracing stuff entirely in favor of KHR? 

There's little reason for anyone to use it for new software, KHR is the only future.

Comment (XAMPPRocky) [reply crates/rustc_codegen_spirv/src/symbols.rs:146]: Maybe? `HitTNV` seems to be the only builtin that wasn't brought over to the KHR extension. So there might be valid uses for it, but I don't know enough about it to be able to say one way or another.

Comment (expenses) [reply crates/rustc_codegen_spirv/src/symbols.rs:146]: Supposedly it's just an alias of a variable that is in KHR:
https://github.com/KhronosGroup/GLSL/blob/ab928c80851b96f3889f1287d4ce0003a9bb86e4/extensions/nv/GLSL_NV_ray_tracing.txt#L657-L658

Comment (hrydgard) [reply crates/rustc_codegen_spirv/src/symbols.rs:146]: Yeah it's nothing special and is only part of the NV extension, which can now be considered deprecated.

Comment (hrydgard) [reply crates/rustc_codegen_spirv/src/symbols.rs:146]: Please just delete it along with the other NV stuff.

File: tests/ui/arch/trace_ray_khr.rs
Comment (hrydgard) [inline tests/ui/arch/trace_ray_khr.rs:5]: wasn't uniform in this position only for uniform buffers?

File: crates/rustc_codegen_spirv/src/symbols.rs
Comment (XAMPPRocky) [reply crates/rustc_codegen_spirv/src/symbols.rs]: Well, I've removed `HitTNV`, I'm going to leave any other pre-existing NV builtins however, as the rest are unrelated to ray tracing, and should be migrated separately as their own PRs.

File: tests/ui/arch/trace_ray_khr.rs
Comment (XAMPPRocky) [reply tests/ui/arch/trace_ray_khr.rs]: Yep, I was just rushing to fix it and didn't type `_constant`, it should be `uniform_constant` but that shouldn't be needed once #567 is merged.

File: crates/rustc_codegen_spirv/Cargo.toml
Comment (khyperia) [inline crates/rustc_codegen_spirv/Cargo.toml:40]: Marking this as request-change just to make sure we don't merge this (it's okay for a draft, but we shouldn't merge it)

File: crates/spirv-std/src/ray_tracing.rs
Comment (khyperia) [inline crates/spirv-std/src/ray_tracing.rs:8]: I'm a little worried about weird edge cases with rustc thinking it's a ZST, I think I'd prefer this to be a `u32` like all the other fake types

**Review by khyperia — CHANGES_REQUESTED** [review]

File: crates/spirv-std/src/ray_tracing.rs
Comment (XAMPPRocky) [reply crates/spirv-std/src/ray_tracing.rs:8]: Shouldn't it be treated as a ZST though? The spec says it has no defined size or bit pattern. I'm fine with changing it though.

Comment (khyperia) [reply crates/spirv-std/src/ray_tracing.rs:8]: "no defined size" is very different than "zero size" - it's the difference between `[T]` and `[T; 0]`. Images are also "no defined size or bit pattern", and we use u32 for those.

File: crates/spirv-std/src/arch/ray_tracing.rs
Comment (khyperia) [inline crates/spirv-std/src/arch/ray_tracing.rs:86]: ... hold on, looking up the docs for this, `ray_origin` should be a straight vector, not a pointer to a vector, yet validation succeeds. I'm scared now, spirv-val probably doesn't validate this instruction at all, and we should probably throw this at an actual implementation to be sure?

Comment (XAMPPRocky) [reply crates/spirv-std/src/arch/ray_tracing.rs:86]: Oh yeah definitely, I don't think there's much validation of ray tracing instructions happening at all.

File: crates/spirv-std/src/arch/ray_tracing.rs
Comment (Jasper-Bekkers) [inline crates/spirv-std/src/arch/ray_tracing.rs:18]: Small nit :-)

Comment (XAMPPRocky) [reply crates/spirv-std/src/arch/ray_tracing.rs:18]: Nit that it's called blah or that it's there at all? The `OpLabel` has to be there because currently we don't correctly handle terminators in inline assembly. https://github.com/EmbarkStudios/rust-gpu/issues/564

Comment (Jasper-Bekkers) [reply crates/spirv-std/src/arch/ray_tracing.rs:18]: Yeah just about the name.

File: crates/spirv-std/src/arch/ray_tracing.rs
Comment (khyperia) [inline crates/spirv-std/src/arch/ray_tracing.rs:29]: If the result is a u8, it can be an `out(reg)`, right? No need to do the default value and `&mut` dance.

File: crates/spirv-std/src/arch/ray_tracing.rs
Comment (khyperia) [inline crates/spirv-std/src/arch/ray_tracing.rs:27]: Result type should be a `%bool` - from the spec:

> Result Type must be a scalar boolean.

File: crates/spirv-std/src/arch/ray_tracing.rs
Comment (khyperia) [inline crates/spirv-std/src/arch/ray_tracing.rs:81]: I don't think there's a test for this function

**Review by khyperia — CHANGES_REQUESTED** [review]

File: crates/spirv-std/src/arch/ray_tracing.rs
Comment (XAMPPRocky) [reply crates/spirv-std/src/arch/ray_tracing.rs:27]: Yeah it should, but it doesn't seem to work. Using bool seems to insert an `OpINotEqual` instruction, which causes it error.

```rust
pub unsafe fn report_intersection(hit: f32, hit_kind: u32) -> bool {
    let result: u8;

    asm! {
        "%bool = OpTypeBool",
        "%u8 = OpTypeInt 8 0",
        "{result} = OpReportIntersectionKHR %bool {hit} {hit_kind}",
        result = out(reg) result,
        hit = in(reg) hit,
        hit_kind = in(reg) hit_kind,
    };

    result != 0
}
```

```spirv
; SPIR-V
; Version: 1.3
; Generator: Google rspirv; 0
; Bound: 17
; Schema: 0
               OpCapability Shader
               OpCapability Int8
               OpCapability RayTracingKHR
               OpExtension "SPV_KHR_ray_tracing"
               OpMemoryModel Logical Simple
               OpEntryPoint IntersectionNV %1 "main"
               OpName %report_intersection_khr__main "report_intersection_khr::main"
       %bool = OpTypeBool
      %float = OpTypeFloat 32
       %uint = OpTypeInt 32 0
      %uchar = OpTypeInt 8 0
    %uchar_0 = OpConstant %uchar 0
       %void = OpTypeVoid
          %9 = OpTypeFunction %void
    %float_2 = OpConstant %float 2
     %uint_4 = OpConstant %uint 4
%report_intersection_khr__main = OpFunction %void None %9
         %12 = OpLabel
         %13 = OpReportIntersectionKHR %bool %float_2 %uint_4
         %14 = OpINotEqual %bool %13 %uchar_0
               OpReturn
               OpFunctionEnd
          %1 = OpFunction %void None %9
         %15 = OpLabel
         %16 = OpFunctionCall %void %report_intersection_khr__main
               OpReturn
               OpFunctionEnd
```

Comment (khyperia) [reply crates/spirv-std/src/arch/ray_tracing.rs]: I'm confused, the spec isn't something you can ignore and make it be a u8 instead, the spec states

> Result Type **must** be a scalar boolean.

maybe I shouldn't have said "should", it _must_ be a boolean.

Comment (khyperia) [reply crates/spirv-std/src/arch/ray_tracing.rs]: This seems to work just fine

```rust
pub unsafe fn report_intersection(hit: f32, hit_kind: u32) -> bool {
    let result: u32;

    asm! {
        "%bool = OpTypeBool",
        "%u32 = OpTypeInt 32 0",
        "%zero = OpConstant %u32 0",
        "%one = OpConstant %u32 0",
        "%result = OpReportIntersectionKHR %bool {hit} {hit_kind}",
        "{result} = OpSelect %u32 %result %one %zero",
        result = out(reg) result,
        hit = in(reg) hit,
        hit_kind = in(reg) hit_kind,
    };

    result != 0
}
```

**Review by khyperia — APPROVED** [review]
LGTM, I'd also like a review from @eddyb

File: crates/rustc_codegen_spirv/src/spirv_type_constraints.rs
Comment (eddyb) [inline crates/rustc_codegen_spirv/src/spirv_type_constraints.rs:730]: Can you change these to be like the other ones which have a comment like this:
```rust
// NOTE(eddyb) we actually use these despite not being in the standard yet.
```
(and the `reserved!` being commented out, not removed)

The reason I want to keep this structure is that the spec still lists these instructions are reserved, and I wanted to be able to track changes to the spec long-term, by mirroring its order and categorization.

File: crates/rustc_codegen_spirv/src/linker/new_structurizer.rs
Comment (eddyb) [inline crates/rustc_codegen_spirv/src/linker/new_structurizer.rs:176]: There's a comment in this file that goes like this:
```rust
    /// The `merge` block must be terminated by one of `OpReturn`, `OpReturnValue`,
    /// `OpKill`, or `OpUnreachable`.
```

Can you update it, or at least make the phrasing less complete (i.e. turn those specific instructions into examples more than an exhaustive list)?

**Review by eddyb — CHANGES_REQUESTED** [review]

**Review by eddyb — APPROVED** [review]

File: crates/rustc_codegen_spirv/src/spirv_type_constraints.rs
Comment (eddyb) [inline crates/rustc_codegen_spirv/src/spirv_type_constraints.rs:759]: Wrong name, should be `SPV_NV_ray_tracing`.

**Review by eddyb — CHANGES_REQUESTED** [review]

File: crates/rustc_codegen_spirv/src/spirv_type_constraints.rs
Comment (XAMPPRocky) [inline crates/rustc_codegen_spirv/src/spirv_type_constraints.rs:759]:
```suggestion
            // reserved!(SPV_NV_ray_tracing)
```

**Review by eddyb — APPROVED** [review]

**Pull request merged**
